### ISSUE TITLE
Store images in same directory as markdown file

### ISSUE THREAD

**Issue by abuisman** [marktext/marktext#2169]
### Describe your feature request

Instead of some global directory I'd like to add images into the same directory as the file I am editting. This makes for the most portable way to work with images since they are right next to my files.

This goes for both pasting and drag-and-dropping.

So:

Drag and drop:
1. I drag and drop an image into Mark Text.
2. Mark Text copies the image to the same directory as the .md file I am editting and adds the reference into the .md file

Pasting:
1. I paste an image from the clipboard:
2. Mark Text stores the image in the same directory as the .md file I am editting and adds the reference into the .md file

Insert url (or unsplash):
1. I paste a url or file-path into the textbox for inserting images
2. Mark Text copies/downloads the image in the same directory as the .md file I am editting and adds the reference into the .md file

**Comment by aisbergde** [comment]
I like this request.

I even want to expand. My idea is:

- I can define a list of possible actions when inserting a picture  
this could be also the current folder, as in the request here, or any other option
- if only option is defined, this option will be used, not asking me (like in current implementation)
- if several possible locations are defined in my settings I will be asked which of the predefined options I want to use: I get all entries from my predefined list and can choose

**Comment by abuisman** [comment]
I have been experimenting in development mode with this and have a rudamentary version of this working. 

(Solvable) issues I have encountered:

- When editting a new document, there is no place for the images yet. So the default location would have to do. On save, you'd have to copy over the images to the place where you saved the file.

- When moving documents, you'd have to move the images as well in order to keep everything correctly organised.

@aisbergde How could we keep your suggestion user friendly though? Add a setting to support this? What is the use-case you have in mind for this? Do you want to handle images differently on a per document basis?

**Comment by aisbergde** [comment]
> A possible solution to that would be to have a per-project settings file, like vscode does with `.vscode/settings.json` that would allow you to tell Mark Text how to handle attachments in the current directory/project.

that's a very good idea! some general options and a way to overwrite this per Workspace, Project, whatever. And that's exactly what I need:
- when I prepare something private, a specific folder
- when it should be compatible with some other aplicaction, a different folder
- blog: public github project
- ...

And this should be a general approach, as we can do in VSC: any settings could be overwritten per workspace, project. 

The question is how to define this project?

Marktext can open folders and there could be optional settings in a folder. And if markdown founds them then they will overwrite the main settings.

**Comment by fxha** [comment]
I think the core issue is a duplicate of #549 but I like the idea with the per directory settings file and also wanted to implement such a feature in sessions (so the settings are saved by directory or project if wanted).

> When editting a new document, there is no place for the images yet. So the default location would have to do. On save, you'd have to copy over the images to the place where you saved the file.

This is currently very tricky but sessions should make this a lot easier as we can copy files in a temporary directory that is handle by the current session and deleted if the file is permanently closed. Unfortunately @Jocs and I  have currently very little time for open-source.

> How could we keep your suggestion user friendly though? Add a setting to support this? What is the use-case you have in mind for this? Do you want to handle images differently on a per document basis?

Additional solution would be #2126.

**Comment by abuisman** [comment]
@fxha yes it looks like having variable-parsing in the image path could be a solution. There'd still have to be features to move images along with a file though.

Also, if you do 'Save as...' the images would have to be copied over as well, etc. Tricky situation would be if the given path (with variables) would result in the same path for two images. You'd have to append '_1` or something to the filename.

> This is currently very tricky but sessions should make this a lot easier as we can copy files in a temporary directory that is handle by the current session and deleted if the file is permanently closed. Unfortunately @Jocs and I have currently very little time for open-source.

I totally get not having time for every feature request that comes your way. I am happy to experiment myself and maybe get something into shape for a PR at some point. Though I have no vue experience, but I can always hobby around for now.

Is there any work done yet on sessions?

I am playing around with implementing the saving-next-to-file feature, could you give me a pointer as to where I can handle the 'save' event of a file in order to move images to the correct place after each save? Also, do I have to parse the markdown file myself to find all image references, or is there an utility somewhere in the codebase already that would let me loop over the images and move them?

I read something about a plugin system, is that already implemented or did I just misunderstand?

**Comment by tatoosh** [comment]
Globale Directory is not nice for Team WIKI Editing.
Images must be in current asset location folder so the mappung must not the Full Path to the image file but only a relativ path like ` ![](assets/2021-03-25-11-42-46-image.png)`

**Comment by Tvax** [comment]
Use `${filename}/../` in `Relative image folder name`:
<img width="657" alt="Screenshot 2024-05-05 at 23 53 57" src="https://github.com/marktext/marktext/assets/19556689/b9cc46cc-e93e-4e90-980e-ed68f7c4188e">

**Comment by Jocs** [comment]
Thanks for the request! This is now supported in MarkText. In **Preferences → Image**, enable **"Prefer relative directory"**, set **"Copy relative to"** to **"Relative to file"**, and leave **"Relative folder name"** blank (or set it to a sub-folder like `assets`). Images inserted by paste, drag-and-drop, or URL will be copied into the same directory as your markdown file (or the named sub-folder). You can also use the `${filename}` variable in the folder name for more advanced layouts (e.g. `${filename}/../` puts images directly alongside the file). Please reopen if you hit a specific case that still doesn't work on the latest version.
